unable to use with Microsoft AD FS server

### Jenkins and plugins versions report

`mvn hpi:run` from master (cddbd8ae4694860c0115cb784f87eaa3f32df2cf)

<details>
  <summary>Environment</summary>

Microsoft AD FS server (2022)  https://learn.microsoft.com/en-us/windows-server/identity/ad-fs/operations/set-up-an-ad-fs-lab-environment#install-the-ad-fs-server-role

  <!-- Paste your environment details below -->
  ```text
nap
  ```

</details>

For whatever reason and despite advertising differently the AD FS server is usign HTTP POST for the return URL.

https://learn.microsoft.com/en-us/windows-server/identity/ad-fs/overview/ad-fs-openid-connect-oauth-flows-scenarios

>  response_mode optional Specifies the method that should be used to send the resulting token back to your app. Defaults to fragment.

This causes issues as the crumb issuer intercepts the call finds there is no crumb and says go away.  

### What Operating System are you using (both controller, and any agents involved in the problem)?

Windows.

### Reproduction steps

1. install an AD FS server as per MSFT "ADSL Lab enironment"
2. create a client for jenkins
3. mvn hpi:run and configure oic plugin
4. attempt to login

### Expected Results

login works

### Actual Results

login fails as the reirect is a crumb

### Anything else?

specifying `response_mode` as `fragement` or `query` which should be the default should force this even if it is supposed to be the default.

### Are you interested in contributing a fix?

_No response_

Note started Jenkins with `-Dhudson.security.csrf.GlobalCrumbIssuerConfiguration.DISABLE_CSRF_PROTECTION=true` (unsafe) for testing and end up with

```
[WARNING] Error while serving http://localhost:8080/jenkins/securityRealm/finishLogin
java.lang.reflect.InvocationTargetException
    at org.kohsuke.stapler.Function$MethodFunction.invoke (Function.java:401)
    at org.kohsuke.stapler.Function$InstanceFunction.invoke (Function.java:409)
    at org.kohsuke.stapler.Function.bindAndInvoke (Function.java:207)
    at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse (Function.java:140)
    at org.kohsuke.stapler.MetaClass$11.doDispatch (MetaClass.java:558)
    at org.kohsuke.stapler.NameBasedDispatcher.dispatch (NameBasedDispatcher.java:59)
    at org.kohsuke.stapler.Stapler.tryInvoke (Stapler.java:770)
    at org.kohsuke.stapler.Stapler.invoke (Stapler.java:900)
    at org.kohsuke.stapler.MetaClass$2.doDispatch (MetaClass.java:224)
    at org.kohsuke.stapler.NameBasedDispatcher.dispatch (NameBasedDispatcher.java:59)
    at org.kohsuke.stapler.Stapler.tryInvoke (Stapler.java:770)
    at org.kohsuke.stapler.Stapler.invoke (Stapler.java:900)
    at org.kohsuke.stapler.Stapler.invoke (Stapler.java:698)
    at org.kohsuke.stapler.Stapler.service (Stapler.java:248)
    at javax.servlet.http.HttpServlet.service (HttpServlet.java:590)
    at org.eclipse.jetty.servlet.ServletHolder.handle (ServletHolder.java:764)
    at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter (ServletHandler.java:1665)
    at hudson.util.PluginServletFilter$1.doFilter (PluginServletFilter.java:163)
    at jenkins.util.HttpServletFilter$1.doFilter (HttpServletFilter.java:76)
    at hudson.util.PluginServletFilter$1.doFilter (PluginServletFilter.java:160)
    at hudson.util.PluginServletFilter.doFilter (PluginServletFilter.java:166)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at jenkins.ErrorAttributeFilter.doFilter (ErrorAttributeFilter.java:29)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at hudson.security.csrf.CrumbFilter.doFilter (CrumbFilter.java:118)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:94)
    at org.jenkinsci.plugins.oic.OicSecurityRealm$1.doFilter (OicSecurityRealm.java:833)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:94)
    at jenkins.security.AcegiSecurityExceptionFilter.doFilter (AcegiSecurityExceptionFilter.java:52)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at hudson.security.UnwrapSecurityExceptionFilter.doFilter (UnwrapSecurityExceptionFilter.java:54)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter (ExceptionTranslationFilter.java:126)
    at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter (ExceptionTranslationFilter.java:120)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter (AnonymousAuthenticationFilter.java:100)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter (RememberMeAuthenticationFilter.java:145)
    at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter (RememberMeAuthenticationFilter.java:101)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter (AbstractAuthenticationProcessingFilter.java:227)
    at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter (AbstractAuthenticationProcessingFilter.java:221)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at jenkins.security.BasicHeaderProcessor.doFilter (BasicHeaderProcessor.java:97)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter (SecurityContextPersistenceFilter.java:117)
    at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter (SecurityContextPersistenceFilter.java:87)
    at hudson.security.HttpSessionContextIntegrationFilter2.doFilter (HttpSessionContextIntegrationFilter2.java:63)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at hudson.security.ChainedServletFilter.doFilter (ChainedServletFilter.java:111)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at hudson.security.ChainedServletFilter.doFilter (ChainedServletFilter.java:111)
    at hudson.security.HudsonFilter.doFilter (HudsonFilter.java:172)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at org.kohsuke.stapler.compression.CompressionFilter.doFilter (CompressionFilter.java:53)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at hudson.util.CharacterEncodingFilter.doFilter (CharacterEncodingFilter.java:86)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter (DiagnosticThreadNameFilter.java:30)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at jenkins.security.SuspiciousRequestFilter.doFilter (SuspiciousRequestFilter.java:38)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at org.eclipse.jetty.servlet.ServletHandler.doHandle (ServletHandler.java:527)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle (ScopedHandler.java:131)
    at org.eclipse.jetty.security.SecurityHandler.handle (SecurityHandler.java:569)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle (HandlerWrapper.java:122)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle (ScopedHandler.java:223)
    at org.eclipse.jetty.server.session.SessionHandler.doHandle (SessionHandler.java:1580)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle (ScopedHandler.java:221)
    at org.eclipse.jetty.server.handler.ContextHandler.doHandle (ContextHandler.java:1384)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextScope (ScopedHandler.java:176)
    at org.eclipse.jetty.servlet.ServletHandler.doScope (ServletHandler.java:484)
    at org.eclipse.jetty.server.session.SessionHandler.doScope (SessionHandler.java:1553)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextScope (ScopedHandler.java:174)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope (ContextHandler.java:1306)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle (ScopedHandler.java:129)
    at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle (ContextHandlerCollection.java:149)
    at org.eclipse.jetty.server.handler.HandlerList.handle (HandlerList.java:51)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle (HandlerWrapper.java:122)
    at org.eclipse.jetty.server.Server.handle (Server.java:563)
    at org.eclipse.jetty.server.HttpChannel$RequestDispatchable.dispatch (HttpChannel.java:1598)
    at org.eclipse.jetty.server.HttpChannel.dispatch (HttpChannel.java:753)
    at org.eclipse.jetty.server.HttpChannel.handle (HttpChannel.java:501)
    at org.eclipse.jetty.server.HttpConnection.onFillable (HttpConnection.java:287)
    at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded (AbstractConnection.java:314)
    at org.eclipse.jetty.io.FillInterest.fillable (FillInterest.java:100)
    at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run (SelectableChannelEndPoint.java:53)
    at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask (AdaptiveExecutionStrategy.java:421)
    at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask (AdaptiveExecutionStrategy.java:390)
    at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce (AdaptiveExecutionStrategy.java:277)
    at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run (AdaptiveExecutionStrategy.java:199)
    at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run (ReservedThreadExecutor.java:411)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob (QueuedThreadPool.java:969)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob (QueuedThreadPool.java:1194)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run (QueuedThreadPool.java:1149)
    at java.lang.Thread.run (Thread.java:829)
Caused by: java.lang.IllegalArgumentException
    at com.google.common.base.Preconditions.checkArgument (Preconditions.java:129)
    at com.google.api.client.util.Preconditions.checkArgument (Preconditions.java:35)
    at com.google.api.client.auth.oauth2.AuthorizationCodeResponseUrl.<init> (AuthorizationCodeResponseUrl.java:90)
    at org.jenkinsci.plugins.oic.OicSession.finishLogin (OicSession.java:192)
    at org.jenkinsci.plugins.oic.OicSecurityRealm.doFinishLogin (OicSecurityRealm.java:1370)
    at java.lang.invoke.MethodHandle.invokeWithArguments (MethodHandle.java:710)
    at org.kohsuke.stapler.Function$MethodFunction.invoke (Function.java:397)
    at org.kohsuke.stapler.Function$InstanceFunction.invoke (Function.java:409)
    at org.kohsuke.stapler.Function.bindAndInvoke (Function.java:207)
    at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse (Function.java:140)
    at org.kohsuke.stapler.MetaClass$11.doDispatch (MetaClass.java:558)
    at org.kohsuke.stapler.NameBasedDispatcher.dispatch (NameBasedDispatcher.java:59)
    at org.kohsuke.stapler.Stapler.tryInvoke (Stapler.java:770)
    at org.kohsuke.stapler.Stapler.invoke (Stapler.java:900)
    at org.kohsuke.stapler.MetaClass$2.doDispatch (MetaClass.java:224)
    at org.kohsuke.stapler.NameBasedDispatcher.dispatch (NameBasedDispatcher.java:59)
    at org.kohsuke.stapler.Stapler.tryInvoke (Stapler.java:770)
    at org.kohsuke.stapler.Stapler.invoke (Stapler.java:900)
    at org.kohsuke.stapler.Stapler.invoke (Stapler.java:698)
    at org.kohsuke.stapler.Stapler.service (Stapler.java:248)
    at javax.servlet.http.HttpServlet.service (HttpServlet.java:590)
    at org.eclipse.jetty.servlet.ServletHolder.handle (ServletHolder.java:764)
    at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter (ServletHandler.java:1665)
    at hudson.util.PluginServletFilter$1.doFilter (PluginServletFilter.java:163)
    at jenkins.util.HttpServletFilter$1.doFilter (HttpServletFilter.java:76)
    at hudson.util.PluginServletFilter$1.doFilter (PluginServletFilter.java:160)
    at hudson.util.PluginServletFilter.doFilter (PluginServletFilter.java:166)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at jenkins.ErrorAttributeFilter.doFilter (ErrorAttributeFilter.java:29)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at hudson.security.csrf.CrumbFilter.doFilter (CrumbFilter.java:118)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:94)
    at org.jenkinsci.plugins.oic.OicSecurityRealm$1.doFilter (OicSecurityRealm.java:833)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:94)
    at jenkins.security.AcegiSecurityExceptionFilter.doFilter (AcegiSecurityExceptionFilter.java:52)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at hudson.security.UnwrapSecurityExceptionFilter.doFilter (UnwrapSecurityExceptionFilter.java:54)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter (ExceptionTranslationFilter.java:126)
    at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter (ExceptionTranslationFilter.java:120)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter (AnonymousAuthenticationFilter.java:100)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter (RememberMeAuthenticationFilter.java:145)
    at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter (RememberMeAuthenticationFilter.java:101)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter (AbstractAuthenticationProcessingFilter.java:227)
    at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter (AbstractAuthenticationProcessingFilter.java:221)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at jenkins.security.BasicHeaderProcessor.doFilter (BasicHeaderProcessor.java:97)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter (SecurityContextPersistenceFilter.java:117)
    at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter (SecurityContextPersistenceFilter.java:87)
    at hudson.security.HttpSessionContextIntegrationFilter2.doFilter (HttpSessionContextIntegrationFilter2.java:63)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at hudson.security.ChainedServletFilter.doFilter (ChainedServletFilter.java:111)
    at hudson.security.ChainedServletFilter$1.doFilter (ChainedServletFilter.java:99)
    at hudson.security.ChainedServletFilter.doFilter (ChainedServletFilter.java:111)
    at hudson.security.HudsonFilter.doFilter (HudsonFilter.java:172)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at org.kohsuke.stapler.compression.CompressionFilter.doFilter (CompressionFilter.java:53)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at hudson.util.CharacterEncodingFilter.doFilter (CharacterEncodingFilter.java:86)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter (DiagnosticThreadNameFilter.java:30)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at jenkins.security.SuspiciousRequestFilter.doFilter (SuspiciousRequestFilter.java:38)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:202)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1635)
    at org.eclipse.jetty.servlet.ServletHandler.doHandle (ServletHandler.java:527)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle (ScopedHandler.java:131)
    at org.eclipse.jetty.security.SecurityHandler.handle (SecurityHandler.java:569)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle (HandlerWrapper.java:122)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle (ScopedHandler.java:223)
    at org.eclipse.jetty.server.session.SessionHandler.doHandle (SessionHandler.java:1580)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle (ScopedHandler.java:221)
    at org.eclipse.jetty.server.handler.ContextHandler.doHandle (ContextHandler.java:1384)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextScope (ScopedHandler.java:176)
    at org.eclipse.jetty.servlet.ServletHandler.doScope (ServletHandler.java:484)
    at org.eclipse.jetty.server.session.SessionHandler.doScope (SessionHandler.java:1553)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextScope (ScopedHandler.java:174)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope (ContextHandler.java:1306)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle (ScopedHandler.java:129)
    at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle (ContextHandlerCollection.java:149)
    at org.eclipse.jetty.server.handler.HandlerList.handle (HandlerList.java:51)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle (HandlerWrapper.java:122)
    at org.eclipse.jetty.server.Server.handle (Server.java:563)
    at org.eclipse.jetty.server.HttpChannel$RequestDispatchable.dispatch (HttpChannel.java:1598)
    at org.eclipse.jetty.server.HttpChannel.dispatch (HttpChannel.java:753)
    at org.eclipse.jetty.server.HttpChannel.handle (HttpChannel.java:501)
    at org.eclipse.jetty.server.HttpConnection.onFillable (HttpConnection.java:287)
    at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded (AbstractConnection.java:314)
    at org.eclipse.jetty.io.FillInterest.fillable (FillInterest.java:100)
    at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run (SelectableChannelEndPoint.java:53)
    at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask (AdaptiveExecutionStrategy.java:421)
    at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask (AdaptiveExecutionStrategy.java:390)
    at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce (AdaptiveExecutionStrategy.java:277)
    at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run (AdaptiveExecutionStrategy.java:199)
    at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run (ReservedThreadExecutor.java:411)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob (QueuedThreadPool.java:969)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob (QueuedThreadPool.java:1194)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run (QueuedThreadPool.java:1149)
    at java.lang.Thread.run (Thread.java:829)
```

--- a/docs/configuration/ADFS.md
+++ b/docs/configuration/ADFS.md
@@ -1,0 +1,26 @@
+# ADFS Provider
+
+[ADFS][1] can be used as as an OpenID Connect identity provider.
+
+## Provider configuration
+
+[This][2] stack overflow step though is a great resource, followed by [This IBM resource][3] for granting the correct permissions.
+
+Where the IBM resource adds 2 individual permissions, 3 are needed and can be performed in one command - e.g.
+`Set-AdfsApplicationPermission -TargetIdentifier fe56f061-c689-45e8-af8d-b8fdf5d1e60f -AddScope 'openid','aza','allatclaims'`
+
+Extra claims (for example users display name) can be added using a similar approach to the groups.
+
+## Plugin configuration
+
+ADFS provides a well known configuration endpoint which can be used for automating endpoint configuration.
+It also supports PKCE verification for additional security.
+
+### User information
+
+Without any extra claims, the user field should be set to `upn`
+
+
+[1]: https://learn.microsoft.com/en-us/windows-server/identity/ad-fs/ad-fs-overview
+[2]: https://stackoverflow.com/questions/55494354/user-groups-as-claims-through-openid-connect-over-adfs/55570286#55570286
+[3]: https://community.ibm.com/community/user/security/blogs/laurent-lapiquionne1/2020/07/21/how-to-configure-igi-service-center-to-authent

--- a/docs/configuration/README.md
+++ b/docs/configuration/README.md
@@ -9,6 +9,7 @@ There are specifics instructions for well known providers:
 
 * [Google Provider](GOOGLE.md)
 * [Gitlab Provider](GITLAB.md)
+* [Microsoft AD FS](ADFS.md)
 
 This page contains the reference of plugin's configuration.
 

--- a/pom.xml
+++ b/pom.xml
@@ -129,6 +129,36 @@
   </pluginRepositories>
 
   <build>
+    <pluginManagement>
+      <plugins>
+        <!--This plugin's configuration is used to store Eclipse m2e settings only. It has no influence on the Maven build itself.-->
+        <plugin>
+          <groupId>org.eclipse.m2e</groupId>
+          <artifactId>lifecycle-mapping</artifactId>
+          <version>1.0.0</version>
+          <configuration>
+            <lifecycleMappingMetadata>
+              <pluginExecutions>
+                <pluginExecution>
+                  <pluginExecutionFilter>
+                    <!-- https://github.com/eclipse-m2e/m2e-core/issues/1291 -->
+                    <groupId>org.apache.maven.plugins</groupId>
+                    <artifactId>maven-checkstyle-plugin</artifactId>
+                    <versionRange>[3.4.0,)</versionRange>
+                    <goals>
+                      <goal>check</goal>
+                    </goals>
+                  </pluginExecutionFilter>
+                  <action>
+                    <ignore />
+                  </action>
+                </pluginExecution>
+              </pluginExecutions>
+            </lifecycleMappingMetadata>
+          </configuration>
+        </plugin>
+      </plugins>
+    </pluginManagement>
     <plugins>
       <plugin>
         <artifactId>maven-release-plugin</artifactId>

--- a/pom.xml
+++ b/pom.xml
@@ -112,6 +112,16 @@
       <version>4.5.14</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-core</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-junit-jupiter</artifactId>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
 
   <repositories>

--- a/src/main/java/org/jenkinsci/plugins/oic/OicCrumbExclusion.java
+++ b/src/main/java/org/jenkinsci/plugins/oic/OicCrumbExclusion.java
@@ -1,0 +1,34 @@
+package org.jenkinsci.plugins.oic;
+
+import hudson.Extension;
+import hudson.security.SecurityRealm;
+import hudson.security.csrf.CrumbExclusion;
+import java.io.IOException;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import jenkins.model.Jenkins;
+
+/**
+ * Crumb exclusion to allow POSTing to {@link OicSecurityRealm#doFinishLogin(org.kohsuke.stapler.StaplerRequest)}
+ */
+@Extension
+public class OicCrumbExclusion extends CrumbExclusion {
+
+    @Override
+    public boolean process(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        Jenkins j = Jenkins.getInstanceOrNull();
+        if (j != null) {
+            SecurityRealm sr = j.getSecurityRealm();
+            if (sr instanceof OicSecurityRealm) {
+                if ("/securityRealm/finishLogin".equals(request.getPathInfo())) {
+                    chain.doFilter(request, response);
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+}

--- a/src/main/java/org/jenkinsci/plugins/oic/OicSession.java
+++ b/src/main/java/org/jenkinsci/plugins/oic/OicSession.java
@@ -41,6 +41,8 @@ import java.security.SecureRandom;
 import java.util.Map;
 import java.util.UUID;
 import javax.servlet.http.HttpSession;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.DoNotUse;
 import org.kohsuke.stapler.HttpRedirect;
 import org.kohsuke.stapler.HttpResponse;
 import org.kohsuke.stapler.Stapler;
@@ -191,19 +193,16 @@ abstract class OicSession implements Serializable {
      * @return an {@link HttpResponse}
      */
     public HttpResponse finishLogin(StaplerRequest request, AuthorizationCodeFlow flow) throws IOException {
-        StringBuffer buf = request.getRequestURL();
+        final String requestURL;
         if (request.getQueryString() != null) {
+            StringBuffer buf = request.getRequestURL();
             buf.append('?').append(request.getQueryString());
+            requestURL = buf.toString();
         } else {
             // some providers ADFS! post data using a form rather than the queryString.
-            Map<String, String[]> parameterMap = request.getParameterMap();
-            UriComponentsBuilder queryBuilder = UriComponentsBuilder.fromHttpUrl(buf.toString());
-            for (Map.Entry<String, String[]> entry : parameterMap.entrySet()) {
-                queryBuilder.queryParam(entry.getKey(), (Object[]) entry.getValue());
-            }
-            buf = new StringBuffer(queryBuilder.build().toUriString());
+            requestURL = convertFormToQueryParameters(request);
         }
-        AuthorizationCodeResponseUrl responseUrl = new AuthorizationCodeResponseUrl(buf.toString());
+        AuthorizationCodeResponseUrl responseUrl = new AuthorizationCodeResponseUrl(requestURL);
         if (!state.equals(responseUrl.getState())) {
             return new Failure("State is invalid");
         }
@@ -226,6 +225,17 @@ abstract class OicSession implements Serializable {
         return onSuccess(code, flow);
     }
 
+    @VisibleForTesting
+    @Restricted(DoNotUse.class)
+    protected static String convertFormToQueryParameters(StaplerRequest request) {
+        Map<String, String[]> parameterMap = request.getParameterMap();
+        UriComponentsBuilder queryBuilder =
+                UriComponentsBuilder.fromHttpUrl(request.getRequestURL().toString());
+        for (Map.Entry<String, String[]> entry : parameterMap.entrySet()) {
+            queryBuilder.queryParam(entry.getKey(), (Object[]) entry.getValue());
+        }
+        return queryBuilder.build().toUriString();
+    }
     /**
      * Where was the user trying to navigate to when they had to login?
      *

--- a/src/main/java/org/jenkinsci/plugins/oic/OicSession.java
+++ b/src/main/java/org/jenkinsci/plugins/oic/OicSession.java
@@ -38,12 +38,14 @@ import java.nio.charset.StandardCharsets;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
 import java.security.SecureRandom;
+import java.util.Map;
 import java.util.UUID;
 import javax.servlet.http.HttpSession;
 import org.kohsuke.stapler.HttpRedirect;
 import org.kohsuke.stapler.HttpResponse;
 import org.kohsuke.stapler.Stapler;
 import org.kohsuke.stapler.StaplerRequest;
+import org.springframework.web.util.UriComponentsBuilder;
 
 import static org.jenkinsci.plugins.oic.OicSecurityRealm.ensureStateAttribute;
 
@@ -192,6 +194,14 @@ abstract class OicSession implements Serializable {
         StringBuffer buf = request.getRequestURL();
         if (request.getQueryString() != null) {
             buf.append('?').append(request.getQueryString());
+        } else {
+            // some providers ADFS! post data using a form rather than the queryString.
+            Map<String, String[]> parameterMap = request.getParameterMap();
+            UriComponentsBuilder queryBuilder = UriComponentsBuilder.fromHttpUrl(buf.toString());
+            for (Map.Entry<String, String[]> entry : parameterMap.entrySet()) {
+                queryBuilder.queryParam(entry.getKey(), (Object[]) entry.getValue());
+            }
+            buf = new StringBuffer(queryBuilder.build().toUriString());
         }
         AuthorizationCodeResponseUrl responseUrl = new AuthorizationCodeResponseUrl(buf.toString());
         if (!state.equals(responseUrl.getState())) {

--- a/src/test/java/org/jenkinsci/plugins/oic/OicCrumbExclusionTest.java
+++ b/src/test/java/org/jenkinsci/plugins/oic/OicCrumbExclusionTest.java
@@ -1,0 +1,101 @@
+package org.jenkinsci.plugins.oic;
+
+import javax.servlet.FilterChain;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import jenkins.model.Jenkins;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.MockedStatic;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.lenient;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class OicCrumbExclusionTest {
+
+    @Mock
+    Jenkins jenkins;
+
+    @Mock
+    MockedStatic<Jenkins> staticJenkins;
+
+    @Mock
+    OicSecurityRealm oicSecurityRealm;
+
+    @Mock
+    FilterChain chain;
+
+    @Mock
+    HttpServletRequest request;
+
+    @Mock
+    HttpServletResponse response;
+
+    private void withJenkins() {
+        staticJenkins.when(Jenkins::getInstanceOrNull).thenReturn(jenkins);
+    }
+
+    private void withRequestPath(String path) {
+        lenient().when(request.getPathInfo()).thenReturn(path);
+    }
+
+    private void withOicSecurityRealm() {
+        when(jenkins.getSecurityRealm()).thenReturn(oicSecurityRealm);
+    }
+
+    @Test
+    void exclusion_applies_when_realm_is_OIC_and_path_is_finishLogin() throws Exception {
+        withJenkins();
+        withRequestPath("/securityRealm/finishLogin");
+        withOicSecurityRealm();
+
+        OicCrumbExclusion oicCrumbExclusion = new OicCrumbExclusion();
+        assertTrue("path should be excluded", oicCrumbExclusion.process(request, response, chain));
+        Mockito.verify(chain, times(1)).doFilter(request, response);
+    }
+
+    @Test
+    void exclusion_does_not_apply_when_realm_is_OIC_and_path_is_not_finishLogin() throws Exception {
+        withJenkins();
+        withRequestPath("/securityRealm/anything");
+        withOicSecurityRealm();
+
+        OicCrumbExclusion oicCrumbExclusion = new OicCrumbExclusion();
+        assertFalse("path should not be excluded", oicCrumbExclusion.process(request, response, chain));
+        Mockito.verify(chain, times(0)).doFilter(request, response);
+    }
+
+    @Test
+    void exclusion_does_not_apply_when_realm_is_not_OIC_and_path_is_finishLogin() throws Exception {
+        withJenkins();
+        withRequestPath("/securityRealm/finishLogin");
+
+        OicCrumbExclusion oicCrumbExclusion = new OicCrumbExclusion();
+        assertFalse("path should not be excluded", oicCrumbExclusion.process(request, response, chain));
+        Mockito.verify(chain, times(0)).doFilter(request, response);
+    }
+
+    @Test
+    void exclusion_does_not_apply_when_realm_is_not_OIC_and_path_is_not_finishLogin() throws Exception {
+        withJenkins();
+        withRequestPath("/securityRealm/anything");
+
+        OicCrumbExclusion oicCrumbExclusion = new OicCrumbExclusion();
+        assertFalse("path should not be excluded", oicCrumbExclusion.process(request, response, chain));
+        Mockito.verify(chain, times(0)).doFilter(request, response);
+    }
+
+    @Test
+    void exclusion_does_not_apply_when_jenkins_is_not_set() throws Exception {
+        OicCrumbExclusion oicCrumbExclusion = new OicCrumbExclusion();
+        assertFalse("path should not be excluded", oicCrumbExclusion.process(request, response, chain));
+        Mockito.verify(chain, times(0)).doFilter(request, response);
+    }
+}

--- a/src/test/java/org/jenkinsci/plugins/oic/OicSessionTest.java
+++ b/src/test/java/org/jenkinsci/plugins/oic/OicSessionTest.java
@@ -2,15 +2,20 @@ package org.jenkinsci.plugins.oic;
 
 import com.google.api.client.auth.oauth2.AuthorizationCodeFlow;
 import java.io.IOException;
+import java.util.SortedMap;
+import java.util.TreeMap;
 import jenkins.model.Jenkins;
-import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
 import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.WithoutJenkins;
 import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.StaplerRequest;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
 public class OicSessionTest {
 
@@ -21,7 +26,6 @@ public class OicSessionTest {
 
     private static final String from = "fromAddy";
 
-    @Before
     public void init() throws IOException {
         TestRealm realm = new TestRealm.Builder("http://localhost/")
                 .WithMinimalDefaults().WithScopes("openid").build();
@@ -44,12 +48,30 @@ public class OicSessionTest {
     }
 
     @Test
-    public void getFrom() {
+    public void getFrom() throws Exception {
+        init();
         assertEquals(from, session.getFrom());
     }
 
     @Test
-    public void getState() {
+    public void getState() throws Exception {
+        init();
         assertNotEquals("", session.getState());
+    }
+
+    @Test
+    @WithoutJenkins
+    public void testFormToQueryParameters() {
+        StaplerRequest sr = mock(StaplerRequest.class);
+        when(sr.getRequestURL())
+                .thenReturn(new StringBuffer("http://domain.invalid/jenkins/securityRealm/finishLogin"));
+        SortedMap<String, String[]> parametersMap = new TreeMap<>();
+        parametersMap.put("param1", new String[] {"p1k1"});
+        parametersMap.put("param2", new String[] {"p2k1", "p2k2"});
+        when(sr.getParameterMap()).thenReturn(parametersMap);
+        String converted = OicSession.convertFormToQueryParameters(sr);
+        assertEquals(
+                "http://domain.invalid/jenkins/securityRealm/finishLogin?param1=p1k1&param2=p2k1&param2=p2k2",
+                converted);
     }
 }

